contracts/modules/staking/StakingProducts.sol

```diff
@@ -0,0 +1,513 @@
+  }
+
+  function getEffectiveWeight(
+    uint poolId,
+    uint productId,
+    uint targetWeight,
+    uint globalCapacityRatio,
+    uint capacityReductionRatio
+  ) public view returns (uint effectiveWeight) {
+
+    IStakingPool stakingPool = getStakingPool(poolId);
+
+    return _getEffectiveWeight(
+      stakingPool,
+      productId,
+      targetWeight,
+      globalCapacityRatio,
+      capacityReductionRatio
+    );
+  }
+
+  function _getEffectiveWeight(
+    IStakingPool stakingPool,
+    uint productId,
+    uint targetWeight,
+    uint globalCapacityRatio,
+    uint capacityReductionRatio
+  ) internal view returns (uint16 effectiveWeight) {
+
+    uint[] memory trancheCapacities = stakingPool.getTrancheCapacities(
+      productId,
+      block.timestamp / TRANCHE_DURATION, // first active tranche id
+      MAX_ACTIVE_TRANCHES,
+      globalCapacityRatio,
+      capacityReductionRatio
+    );
+
+    uint totalCapacity = Math.sum(trancheCapacities);
+
+    if (totalCapacity == 0) {
+      return targetWeight.toUint16();
+    }
+
+    uint[] memory activeAllocations = stakingPool.getActiveAllocations(productId);
+    uint totalAllocation = Math.sum(activeAllocations);
+    uint actualWeight = Math.min(totalAllocation * WEIGHT_DENOMINATOR / totalCapacity, type(uint16).max);
+
+    return Math.max(targetWeight, actualWeight).toUint16();
+  }
+
+  function setInitialProducts(uint poolId, ProductInitializationParams[] memory params) public onlyInternal {
+
+    uint totalTargetWeight;
+
+    for (uint i = 0; i < params.length; i++) {
+
+      ProductInitializationParams memory param = params[i];
+
+      if (param.targetPrice > TARGET_PRICE_DENOMINATOR) {
+        revert TargetPriceTooHigh();
+      }
+
+      if (param.weight > WEIGHT_DENOMINATOR) {
+        revert TargetWeightTooHigh();
+      }
+
+      StakedProduct memory product;
+      product.bumpedPrice = param.initialPrice;
+      product.bumpedPriceUpdateTime = block.timestamp.toUint32();
+      product.targetPrice = param.targetPrice;
+      product.targetWeight = param.weight;
+
+      // sstore
+      _products[poolId][param.productId] = product;
+
+      totalTargetWeight += param.weight;
+    }
+
+    if (totalTargetWeight > MAX_TOTAL_WEIGHT) {
+      revert TotalTargetWeightExceeded();
+    }
+
+    weights[poolId] = Weights({
+      totalTargetWeight: totalTargetWeight.toUint32(),
+      totalEffectiveWeight: 0
+    });
+  }
+
+  /* pricing code */
+  function getPremium(
+    uint poolId,
+    uint productId,
+    uint period,
+    uint coverAmount,
+    uint initialCapacityUsed,
+    uint totalCapacity,
+    uint globalMinPrice,
+    bool useFixedPrice,
+    uint nxmPerAllocationUnit,
+    uint allocationUnitsPerNXM
+  ) public returns (uint premium) {
+
+    // TODO: WARNING! EXTREMELY IMPORTANT!
+    // TODO: uncomment this after moving the StakingProducts tests to a separate folder
+    // TODO: need to use a SPMockStakingProducts when testing StakingPool
```

Comment:
stating the obvious but yes still left to do

Is this covered as part of this PR? Or separate?

Reply:
[#702](https://github.com/NexusMutual/smart-contracts/issues/702)